nasl/nasl_packet_forgery.c

```diff
@@ -818,6 +847,269 @@ set_tcp_elements (lex_ctxt *lexic)
+          opt_sack_perm->len = TCPOLEN_SACK_PERMITTED;
+
+          // Need reallocated memory because options requires it.
+          if (total_opt_len > opt_size_allocated)
+            {
+              opt_size_allocated = ((total_opt_len / 4) + 1) * 4;
+              opts = g_realloc (opts, sizeof (char) * opt_size_allocated);
+              ptr_opts_pos = opts + current_opt_len;
+            }
+
+          memcpy (ptr_opts_pos, (u_char *) opt_sack_perm,
+                  sizeof (struct tcp_opt_sack_perm));
+          ptr_opts_pos = ptr_opts_pos + sizeof (struct tcp_opt_sack_perm);
+          g_free (opt_sack_perm);
+          break;
+        case TCPOPT_TIMESTAMP:
+          tcp_opt_val = get_int_var_by_num (lexic, i + 1, -1);
+          tcp_opt_val2 = get_int_var_by_num (lexic, i + 2, -1);
+          i = i + 2;
+          if (tcp_opt_val < 0)
+            nasl_perror (lexic, "%s: Invalid value for TCP option Timestamp'\n",
+                         __func__);
+          opt_tstamp = g_malloc0 (sizeof (struct tcp_opt_tstamp));
+          total_opt_len += TCPOLEN_TIMESTAMP;
+          opt_tstamp->kind = TCPOPT_TIMESTAMP;
+          opt_tstamp->len = TCPOLEN_TIMESTAMP;
+          opt_tstamp->tstamp = FIX (tcp_opt_val);
+          opt_tstamp->e_tstamp = FIX (tcp_opt_val2);
+
+          // Need reallocated memory because options requires it.
+          if (total_opt_len > opt_size_allocated)
+            {
+              opt_size_allocated = ((total_opt_len / 4) + 1) * 4;
+              opts = g_realloc (opts, sizeof (char) * opt_size_allocated);
+              ptr_opts_pos = opts + current_opt_len;
+            }
+
+          memcpy (ptr_opts_pos, (u_char *) opt_tstamp,
+                  sizeof (struct tcp_opt_tstamp));
+          ptr_opts_pos = ptr_opts_pos + sizeof (struct tcp_opt_tstamp);
+          g_free (opt_tstamp);
+          break;
+        case TCPOPT_NOP:
+        case TCPOPT_EOL:
+        case TCPOPT_SACK: /* Experimental, not supported */
+        default:
+          nasl_perror (lexic, "%s: TCP option %d not supported\n", __func__,
+                       tcp_opt);
+          break;
+        }
+    }
+
+  // Add NOP padding and End Of Option list kinds.
+  current_opt_len = total_opt_len;
+  eol = TCPOPT_EOL;
+  nop = TCPOPT_NOP;
+  if (total_opt_len % 4 == 0)
+    {
+      opt_size_allocated = opt_size_allocated + 4;
+      opts = g_realloc (opts, sizeof (char) * opt_size_allocated);
+      ptr_opts_pos = opts + total_opt_len;
+    }
+  if (current_opt_len < opt_size_allocated - 1)
+    {
+      // Add NOPs
+      for (i = current_opt_len; i < opt_size_allocated - 1; i++)
+        {
+          memcpy (ptr_opts_pos, &nop, 1);
+          total_opt_len++;
+          ptr_opts_pos++;
+        }
+    }
+  // Add EOL
+  memcpy (ptr_opts_pos, &eol, 1);
+
+  if (ip->ip_hl * 4 > pktsz)
+    tcp =
+      (struct tcphdr *) (pkt
+                         + 20); /* ip->ip_hl is bogus, we work around that */
```

Comment:
The formatting is bad here. Maybe put the comment above the line so the formatter does not do this.